src/main/scala/com/advancedtelematic/campaigner/db/Campaigns.scala

```diff
@@ -135,14 +128,31 @@ protected [db] class Campaigns(implicit db: Database, ec: ExecutionContext)
 
   def findClientCampaign(campaignId: CampaignId): Future[GetCampaign] = for {
     c <- campaignRepo.find(campaignId)
-    retryIds <- campaignRepo.findRetryCampaignIdsOf(campaignId)
+    retryIds <- campaignRepo.findRetryCampaignsOf(campaignId).map(_.map(_.id))
     groups <- db.run(findGroupsAction(c.id))
     metadata <- campaignMetadataRepo.findFor(campaignId)
   } yield GetCampaign(c, retryIds, groups, metadata)
 
   def findCampaignsByUpdate(update: UpdateId): Future[Seq[Campaign]] =
     db.run(campaignRepo.findByUpdateAction(update))
 
+  /**
+   * Given a set of campaign IDs, finds all device updates that happened in
+   * these campaigns, selects the most recent failures and returns them.
+   */
+  protected[db] def findFailedDeviceUpdatesAction(campaignIds: Set[CampaignId]): DBIO[Set[DeviceUpdate]] = {
+    Schema.deviceUpdates
+      .join(Schema.deviceUpdates
+      .filter(upd => upd.campaignId.inSet(campaignIds))
+        .groupBy(_.deviceId)
+        .map { case (id, upd) => (id, upd.map(_.updatedAt).max) })
+        .on { (fst, snd) => fst.deviceId === snd._1 && fst.updatedAt === snd._2 }
```

Comment:
In plain SQL I'd use `rank()` over partition by `device_id` sorted by `update_at`, but I don't know how to express it in Slick DSL. Doing it in `sql`-escaped string would require duplicating (de)serialization in terms of `GetResult/SetParameter` implicit objects. Taking into account both of these things, I thought that we'll be fine with a `join` for now, and explore other options when the need arise. I'd appreciate your help if you know an alternative that's more efficient and elegant.

_Side note:_ I think that calculating stats upon a request is quite expensive when the numbers of campaigns and devices are sufficiently large. I'd do it at the time of receiving a device installation report and persist in a separate table, so that a result could be returned using simple `select`. Again, this measure would only be necessary when we have perf. issues with the current impl.